sike/sike.go

```diff
@@ -0,0 +1,219 @@
+// PKE interface
+//
+
+// Uses SIKE public key to encrypt plaintext. Requires cryptographically secure PRNG
+// Returns ciphertext in case encryption succeeds. Returns error in case PRNG fails
+// or wrongly formated input was provided.
+func Encrypt(rng io.Reader, pub *PublicKey, ptext []byte) ([]byte, error) {
+	var params = pub.Params()
+	var ptextLen = uint(len(ptext))
+	// c1 must be security level + 64 bits (see [SIKE] 1.4 and 4.3.3)
+	if ptextLen != (params.KemSize + 8) {
+		return nil, errors.New("Unsupported message length")
+	}
+
+	skA := NewPrivateKey(params.Id, KeyVariant_SIDH_A)
+	err := skA.Generate(rng)
+	if err != nil {
+		return nil, err
+	}
+
+	pkA, _ := GeneratePublicKey(skA) // Never fails
+	return encrypt(skA, pkA, pub, ptext)
+}
+
+// Uses SIKE private key to decrypt ciphertext. Returns plaintext in case
+// decryption succeeds or error in case unexptected input was provided.
+// Constant time
+func Decrypt(prv *PrivateKey, ctext []byte) ([]byte, error) {
+	var params = prv.Params()
+	var n [40]byte // n can is max 320-bit (see 1.4 of [SIKE])
+	var c1_len int
+	var pk_len = params.PublicKeySize
+
+	if prv.Variant() != KeyVariant_SIKE {
+		return nil, errors.New("wrong key type")
+	}
+
+	// ctext is a concatenation of (pubkey_A || c1=ciphertext)
+	// it must be security level + 64 bits (see [SIKE] 1.4 and 4.3.3)
+	c1_len = len(ctext) - pk_len
+	if c1_len != (int(params.KemSize) + 8) {
+		return nil, errors.New("wrong size of cipher text")
+	}
+
+	c0 := NewPublicKey(params.Id, KeyVariant_SIDH_A)
+	err := c0.Import(ctext[:pk_len])
+	if err != nil {
+		return nil, err
+	}
+	j, err := DeriveSecret(prv, c0)
+	if err != nil {
+		return nil, err
+	}
+
+	cshakeSum(n[:c1_len], j, F)
+	for i, _ := range n[:c1_len] {
+		n[i] ^= ctext[pk_len+i]
+	}
+
+	return n[:c1_len], nil
+}
+
+// -----------------------------------------------------------------------------
+// KEM interface
+//
+
+// Encapsulation receives the public key and generates SIKE ciphertext and shared secret.
+// The generated ciphertext is used for authentication.
+// The rng must be cryptographically secure PRNG.
+// Error is returned in case PRNG fails or wrongly formated input was provided.
+func Encapsulate(rng io.Reader, pub *PublicKey) (ctext []byte, secret []byte, err error) {
+	var params = pub.Params()
+	// Buffer for random, secret message
+	var ptext = make([]byte, params.MsgLen)
+	// r = G(ptext||pub)
+	var r = make([]byte, params.SecretKeySize)
+	// Resulting shared secret
+	secret = make([]byte, params.KemSize)
+
+	// Generate ephemeral value
+	_, err = io.ReadFull(rng, ptext)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	h := cshake.NewCShake256(nil, G)
+	h.Write(ptext)
+	h.Write(pub.Export())
+	h.Read(r)
+
+	// cSHAKE256 implementation is byte oriented. Ensure bitlength is less then to E2
+	r[len(r)-1] &= params.A.MaskBytes[0]
+	r[len(r)-2] &= params.A.MaskBytes[1] // clear high bits, so scalar < 2*732
+
+	// (c0 || c1) = Enc(pkA, ptext; r)
```

Comment:
Why not use the Encrypt function defined above? Seems like the function could be modified to allow you to pass in the desired randomness. This isn't necessary, but seems like it could avoid some code reuse.

Reply:
Encrypt() works differently than private encrypt() function. First of all it is an interface for SIKE CPA-secure PKE, second - it generates SIDH_A key pair which isn't required here.